### PR TITLE
Npm: Ignore bundled sub-dependencies

### PR DESCRIPTION
**Pull request by feelepxyz**
Thinking that we shouldn't update bundled sub-dependencies as they have
been bundled into the release at an exact version by a parent using
`bundledDependencies`.

For example, fsevents < 2 bundles node-pre-gyp meaning all it's
sub-dependnecies get bundled into the release tarball at publish time so
you always get the same sub-dependency versions if you re-install a
specific version of fsevents.

Updating the sub-dependency (by deleting the entry) does seem to work
but it gets removed from the bundled set of dependencies and moved top
level resulting in a bunch of package duplication which is pretty
confusing.

To fix this properly we would need to try and update the parent but
this gets a bit crazy if we need to go up several levels which is the
case when updating `tar` here: `chokidar > fsevents > node-pre-gyp >
tar`.

### CODE REVIEW

**Comment by feelepxyz** [comment]
@greysteil thoughts on this? Thinking this might be the best thing to do. Seems super rare that you have bundled dependencies and fsevents moved away from it in version 2.

File: npm_and_yarn/.byebug_history [hunk 1]
```diff
@@ -0,0 +1,2 @@
+exit
+dependency.extra&.fetch(:bundled)
```
Comment (greysteil) [inline npm_and_yarn/.byebug_history]: ✂️ and ignore this file?

**Comment by feelepxyz** [comment]
@greysteil would love your 👀 on this when you get a moment!

File: common/lib/dependabot/dependency.rb [hunk 2]
```diff
@@ -27,17 +27,19 @@ def self.register_display_name_builder(package_manager, name_builder)
...
+                   previous_version: nil, previous_requirements: nil,
+                   extra: nil)
       @name = name
       @version = version
       @requirements = requirements.map { |req| symbolize_keys(req) }
       @previous_version = previous_version
       @previous_requirements =
         previous_requirements&.map { |req| symbolize_keys(req) }
       @package_manager = package_manager
+      @extra = extra
```
Comment (greysteil) [inline common/lib/dependabot/dependency.rb]: Maybe `metadata` over `extra`?

File: npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/subdependency_version_resolver.rb [hunk 3]
```diff
@@ -214,6 +215,26 @@ def package_files
...
+        # For example, fsevents < 2 bundles node-pre-gyp meaning all it's
+        # sub-dependnecies get bundled into the release tarball at publish time
+        # so you always get the same sub-dependency versions if you re-install a
+        # specific version of fsevents.
+        #
+        # Updating the sub-dependency by deleting the entry works but it gets
+        # removed from the bundled set of dependencies and moved top level
+        # resulting in a bunch of package duplication which is pretty confusing.
+        def bundled_dependency?
+          dependency.extra&.fetch(:bundled)
```
Comment (greysteil) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/subdependency_version_resolver.rb]: `dependency.extra&.fetch(:bundled, false) || false`

File: common/lib/dependabot/dependency.rb (hunk 2 above)
Comment (greysteil) [reply common/lib/dependabot/dependency.rb]: Or `subdependency_metadata` and only allow it for cases where `requirements` is an empty array?

Comment (feelepxyz) [reply common/lib/dependabot/dependency.rb]: Yeah that's nicer 👍

File: common/lib/dependabot/dependency.rb [hunk 4]
```diff
@@ -53,7 +55,8 @@ def to_h
         "requirements" => requirements,
         "previous_version" => previous_version,
         "previous_requirements" => previous_requirements,
-        "package_manager" => package_manager
+        "package_manager" => package_manager,
+        "metadata" => metadata
       }
```
Comment (greysteil) [inline common/lib/dependabot/dependency.rb]: Might want to call `compact` here, so that setups without metadata don't include that key.

**Comment by feelepxyz** [comment]
@greysteil pushed up the rename

**Pull request merged**